From ec11b945d4ab369e0fe96aa0cc2dbfabfda30abf Mon Sep 17 00:00:00 2001
From: John Weber <john.weber@technexion.com>
Date: Fri, 2 Feb 2024 15:10:32 -0600
Subject: [PATCH] scripts:setup_cameras: Add support for TEVI and TEVS cameras

---
 scripts/setup_cameras.sh | 201 +++++++++++++++++++++++++++++++++++++++
 1 file changed, 201 insertions(+)

diff --git a/scripts/setup_cameras.sh b/scripts/setup_cameras.sh
index 9c8dba8..94095bc 100755
--- a/scripts/setup_cameras.sh
+++ b/scripts/setup_cameras.sh
@@ -42,6 +42,10 @@ setup_routes(){
 
     OV2312_CAM_FMT='[fmt:SBGGI10_1X10/1600x1300 field: none]'
     IMX390_CAM_FMT='[fmt:SRGGB12_1X12/1936x1100 field: none]'
+    TEVI_AP1302_FMT='[fmt:UYVY8_1X16/1280x720 field: none]'
+    TEVS_FMT='[fmt:UYVY8_1X16/1280x720 field: none]'
+
+    echo "Setting up routes..."
 
     for i in "${!ALL_UB960_FMT_STR[@]}"
     do
@@ -65,6 +69,20 @@ setup_routes(){
             media-ctl -d $id -V "'$UB960_NAME':$UB960_PAD $IMX390_CAM_FMT"
         done
 
+        for name in `media-ctl -d $id -p | grep entity | grep "tevi-ap1302" | cut -d ' ' -f 5`; do
+            UB953_NAME=`media-ctl -d $id -p -e "tevi-ap1302 $name" | grep ub953 | cut -d "\"" -f 2`
+            UB960_NAME=`media-ctl -d $id -p -e "$UB953_NAME" | grep ub960 | cut -d "\"" -f 2`
+            UB960_PAD=`media-ctl -d $id -p -e "$UB953_NAME" | grep ub960 | cut -d : -f 2 | awk '{print $1}'`
+            media-ctl -d $id -V "'$UB960_NAME':$UB960_PAD $TEVI_AP1302_FMT"
+        done
+
+        for name in `media-ctl -d $id -p | grep entity | grep "tevs" | cut -d ' ' -f 5`; do
+            UB953_NAME=`media-ctl -d $id -p -e "tevs $name" | grep ub953 | cut -d "\"" -f 2`
+            UB960_NAME=`media-ctl -d $id -p -e "$UB953_NAME" | grep ub960 | cut -d "\"" -f 2`
+            UB960_PAD=`media-ctl -d $id -p -e "$UB953_NAME" | grep ub960 | cut -d : -f 2 | awk '{print $1}'`
+            media-ctl -d $id -V "'$UB960_NAME':$UB960_PAD $TEVS_FMT"
+        done
+
     done
 
     # CDNS ROUTING
@@ -94,6 +112,25 @@ setup_routes(){
             CSI_BRIDGE_NAME=`media-ctl -d $id -p -e "$UB960_NAME" | grep csi-bridge | cut -d "\"" -f 2`
             media-ctl -d $id -V "'$CSI_BRIDGE_NAME':0/$CSI_PAD $IMX390_CAM_FMT"
         done
+
+        for name in `media-ctl -d $id -p | grep entity | grep "tevi-ap1302" | cut -d ' ' -f 5`; do
+            UB953_NAME=`media-ctl -d $id -p -e "tevi-ap1302 $name" | grep ub953 | cut -d "\"" -f 2`
+            UB960_NAME=`media-ctl -d $id -p -e "$UB953_NAME" | grep ub960 | cut -d "\"" -f 2`
+            UB960_PAD=`media-ctl -d $id -p -e "$UB953_NAME" | grep ub960 | cut -d : -f 2 | awk '{print $1}'`
+            CSI_PAD=`media-ctl -d $id -p -e "$UB960_NAME" | grep $UB960_PAD/.*[ACTIVE] | cut -d "/" -f 3 | awk '{print $1}'`
+            CSI_BRIDGE_NAME=`media-ctl -d $id -p -e "$UB960_NAME" | grep csi-bridge | cut -d "\"" -f 2`
+            media-ctl -d $id -V "'$CSI_BRIDGE_NAME':0/$CSI_PAD $TEVI_AP1302_FMT"
+        done
+
+        for name in `media-ctl -d $id -p | grep entity | grep "tevs" | cut -d ' ' -f 5`; do
+            UB953_NAME=`media-ctl -d $id -p -e "tevs $name" | grep ub953 | cut -d "\"" -f 2`
+            UB960_NAME=`media-ctl -d $id -p -e "$UB953_NAME" | grep ub960 | cut -d "\"" -f 2`
+            UB960_PAD=`media-ctl -d $id -p -e "$UB953_NAME" | grep ub960 | cut -d : -f 2 | awk '{print $1}'`
+            CSI_PAD=`media-ctl -d $id -p -e "$UB960_NAME" | grep $UB960_PAD/.*[ACTIVE] | cut -d "/" -f 3 | awk '{print $1}'`
+            CSI_BRIDGE_NAME=`media-ctl -d $id -p -e "$UB960_NAME" | grep csi-bridge | cut -d "\"" -f 2`
+            media-ctl -d $id -V "'$CSI_BRIDGE_NAME':0/$CSI_PAD $TEVS_FMT"
+        done
+
     done
 
     # CSI2RX ROUTING
@@ -326,6 +363,168 @@ setup_ov5640(){
     done
 }
 
+setup_tevi_ap1302(){
+    TEVI_AP1302_CAM_FMT='[fmt:UYVY8_1X16/1280x720@1/30 field:none]'
+    TEVI_AP1302_STREAM_FMT='[fmt:UYVY8_1X16/1280x720 field:none]'              
+
+    i=0
+    for media_id in {0..1}; do
+    # UB953 FORMATS
+    UB960_FMT_STR=""
+    CDNS_FMT_STR=""
+    CSI2RX_FMT_STR=""
+    for name in `media-ctl -d $media_id -p | grep entity | grep "tevi-ap1302" | cut -d ' ' -f 5`; do
+
+        CAM_SUBDEV=`media-ctl -d $media_id -p -e "tevi-ap1302 $name" | grep v4l-subdev | awk '{print $4}'`
+        
+        UB953_NAME=`media-ctl -d $media_id -p -e "tevi-ap1302 $name" | grep ub953 | cut -d "\"" -f 2`
+        
+        media-ctl -d $media_id --set-v4l2 "'tevi-ap1302 $name':0 $TEVI_AP1302_CAM_FMT"
+        media-ctl -d $media_id -V "'$UB953_NAME':0 $TEVI_AP1302_STREAM_FMT"
+
+        UB960_NAME=`media-ctl -d $media_id -p -e "$UB953_NAME" | grep ub960 | cut -d "\"" -f 2`
+        UB960_PAD=`media-ctl -d $media_id -p -e "$UB953_NAME" | grep ub960 | cut -d : -f 2 | awk '{print $1}'`
+
+        CSI_BRIDGE_NAME=`media-ctl -d $media_id -p -e "$UB960_NAME" | grep csi-bridge | cut -d "\"" -f 2`
+
+        CSI2RX_NAME=`media-ctl -d $media_id -p -e "$CSI_BRIDGE_NAME" | grep "ticsi2rx\"" | cut -d "\"" -f 2`
+
+        LAST_PAD=`echo ${ALL_UB960_FMT_STR[$media_id,$UB960_NAME]} | rev | cut -d'/' -f 1 | rev`
+        LAST_PAD=${LAST_PAD:0:1}
+        if [[ "$LAST_PAD" == "" ]] ; then
+            NEXT_PAD=$UB960_PAD
+        else
+            NEXT_PAD=$(($LAST_PAD+1))
+        fi
+
+        CSI2RX_CONTEXT_NAME="$CSI2RX_NAME context $((NEXT_PAD+1))"
+
+        UB960_FMT_STR="${UB960_PAD}/0 -> 4/$(($NEXT_PAD)) [1]"
+        CDNS_FMT_STR="0/${NEXT_PAD} -> 1/$(($NEXT_PAD)) [1]"
+        CSI2RX_FMT_STR="0/${NEXT_PAD} -> $(($NEXT_PAD+2))/0 [1]"
+
+        # Append UB960 Routes
+        if [[ -v "ALL_UB960_FMT_STR[$media_id,$UB960_NAME]" ]] ; then
+            ALL_UB960_FMT_STR[$media_id,$UB960_NAME]="${ALL_UB960_FMT_STR[$media_id,$UB960_NAME]}, $UB960_FMT_STR"
+        else
+            ALL_UB960_FMT_STR[$media_id,$UB960_NAME]="$UB960_FMT_STR"
+        fi
+
+        # Append CDNS Routes
+        if [[ -v "ALL_CDNS_FMT_STR[$media_id,$CSI_BRIDGE_NAME]" ]] ; then
+            ALL_CDNS_FMT_STR[$media_id,$CSI_BRIDGE_NAME]="${ALL_CDNS_FMT_STR[$media_id,$CSI_BRIDGE_NAME]}, $CDNS_FMT_STR"
+        else
+            ALL_CDNS_FMT_STR[$media_id,$CSI_BRIDGE_NAME]="$CDNS_FMT_STR"
+        fi
+
+        # Append CSIRX Routes
+        if [[ -v "ALL_CSI2RX_FMT_STR[$media_id,$CSI2RX_NAME]" ]] ; then
+            ALL_CSI2RX_FMT_STR[$media_id,$CSI2RX_NAME]="${ALL_CSI2RX_FMT_STR[$media_id,$CSI2RX_NAME]}, $CSI2RX_FMT_STR"
+        else
+            ALL_CSI2RX_FMT_STR[$media_id,$CSI2RX_NAME]="$CSI2RX_FMT_STR"
+        fi
+        CAM_DEV=`media-ctl -d $media_id -p -e "$CSI2RX_CONTEXT_NAME" | grep video | awk '{print $4}'`
+        CAM_DEV_NAME=/dev/video-ap1302-cam$i
+
+        CAM_SUBDEV_NAME=/dev/v4l-ap1302-subdev$i
+
+        ln -snf $CAM_DEV $CAM_DEV_NAME
+        ln -snf $CAM_SUBDEV $CAM_SUBDEV_NAME
+
+        echo -e "${GREEN}TEVI-AP1302 Camera $i detected${NOCOLOR}"
+        echo "    device = $CAM_DEV_NAME"
+        echo "    name = tevi-ap1302"
+        echo "    format = $TEVI_AP1302_CAM_FMT"
+        echo "    subdev_id = $CAM_SUBDEV_NAME"
+        echo "    isp_required = no"
+        echo "    ldc_required = no"
+
+        ((i++))
+    done
+    done
+}
+
+setup_tevs(){
+    TEVS_CAM_FMT='[fmt:UYVY8_1X16/1280x720@1/30 field:none]'
+    TEVS_STREAM_FMT='[fmt:UYVY8_1X16/1280x720 field:none]'              
+
+    i=0
+    for media_id in {0..1}; do
+    # UB953 FORMATS
+    UB960_FMT_STR=""
+    CDNS_FMT_STR=""
+    CSI2RX_FMT_STR=""
+    for name in `media-ctl -d $media_id -p | grep entity | grep "tevs" | cut -d ' ' -f 5`; do
+
+        CAM_SUBDEV=`media-ctl -d $media_id -p -e "tevs $name" | grep v4l-subdev | awk '{print $4}'`
+        
+        UB953_NAME=`media-ctl -d $media_id -p -e "tevs $name" | grep ub953 | cut -d "\"" -f 2`
+        
+        media-ctl -d $media_id --set-v4l2 "'tevs $name':0 $TEVS_CAM_FMT"
+        media-ctl -d $media_id -V "'$UB953_NAME':0 $TEVS_STREAM_FMT"
+
+        UB960_NAME=`media-ctl -d $media_id -p -e "$UB953_NAME" | grep ub960 | cut -d "\"" -f 2`
+        UB960_PAD=`media-ctl -d $media_id -p -e "$UB953_NAME" | grep ub960 | cut -d : -f 2 | awk '{print $1}'`
+
+        CSI_BRIDGE_NAME=`media-ctl -d $media_id -p -e "$UB960_NAME" | grep csi-bridge | cut -d "\"" -f 2`
+
+        CSI2RX_NAME=`media-ctl -d $media_id -p -e "$CSI_BRIDGE_NAME" | grep "ticsi2rx\"" | cut -d "\"" -f 2`
+
+        LAST_PAD=`echo ${ALL_UB960_FMT_STR[$media_id,$UB960_NAME]} | rev | cut -d'/' -f 1 | rev`
+        LAST_PAD=${LAST_PAD:0:1}
+        if [[ "$LAST_PAD" == "" ]] ; then
+            NEXT_PAD=$UB960_PAD
+        else
+            NEXT_PAD=$(($LAST_PAD+1))
+        fi
+
+        CSI2RX_CONTEXT_NAME="$CSI2RX_NAME context $((NEXT_PAD+1))"
+
+        UB960_FMT_STR="${UB960_PAD}/0 -> 4/$(($NEXT_PAD)) [1]"
+        CDNS_FMT_STR="0/${NEXT_PAD} -> 1/$(($NEXT_PAD)) [1]"
+        CSI2RX_FMT_STR="0/${NEXT_PAD} -> $(($NEXT_PAD+2))/0 [1]"
+
+        # Append UB960 Routes
+        if [[ -v "ALL_UB960_FMT_STR[$media_id,$UB960_NAME]" ]] ; then
+            ALL_UB960_FMT_STR[$media_id,$UB960_NAME]="${ALL_UB960_FMT_STR[$media_id,$UB960_NAME]}, $UB960_FMT_STR"
+        else
+            ALL_UB960_FMT_STR[$media_id,$UB960_NAME]="$UB960_FMT_STR"
+        fi
+
+        # Append CDNS Routes
+        if [[ -v "ALL_CDNS_FMT_STR[$media_id,$CSI_BRIDGE_NAME]" ]] ; then
+            ALL_CDNS_FMT_STR[$media_id,$CSI_BRIDGE_NAME]="${ALL_CDNS_FMT_STR[$media_id,$CSI_BRIDGE_NAME]}, $CDNS_FMT_STR"
+        else
+            ALL_CDNS_FMT_STR[$media_id,$CSI_BRIDGE_NAME]="$CDNS_FMT_STR"
+        fi
+
+        # Append CSIRX Routes
+        if [[ -v "ALL_CSI2RX_FMT_STR[$media_id,$CSI2RX_NAME]" ]] ; then
+            ALL_CSI2RX_FMT_STR[$media_id,$CSI2RX_NAME]="${ALL_CSI2RX_FMT_STR[$media_id,$CSI2RX_NAME]}, $CSI2RX_FMT_STR"
+        else
+            ALL_CSI2RX_FMT_STR[$media_id,$CSI2RX_NAME]="$CSI2RX_FMT_STR"
+        fi
+        CAM_DEV=`media-ctl -d $media_id -p -e "$CSI2RX_CONTEXT_NAME" | grep video | awk '{print $4}'`
+        CAM_DEV_NAME=/dev/video-tevs-cam$i
+
+        CAM_SUBDEV_NAME=/dev/v4l-tevs-subdev$i
+
+        ln -snf $CAM_DEV $CAM_DEV_NAME
+        ln -snf $CAM_SUBDEV $CAM_SUBDEV_NAME
+
+        echo -e "${GREEN}TEVS Camera $i detected${NOCOLOR}"
+        echo "    device = $CAM_DEV_NAME"
+        echo "    name = tevs"
+        echo "    format = $TEVS_CAM_FMT"
+        echo "    subdev_id = $CAM_SUBDEV_NAME"
+        echo "    isp_required = no"
+        echo "    ldc_required = no"
+
+        ((i++))
+    done
+    done
+}
+
 setup_USB_camera(){
     ls /dev/v4l/by-path/*usb*video-index0 > /dev/null 2>&1
     if [ "$?" == "0" ]; then
@@ -346,6 +545,8 @@ setup_USB_camera(){
 
 setup_USB_camera
 setup_imx219
+setup_tevi_ap1302
+setup_tevs
 setup_ov5640
 setup_ov2312
 setup_imx390
-- 
2.34.1

